File: apps/api/app/domains/report.py
"""Report contracts, graph snapshot, and in-memory repository.

The current ResearchRun HTTP contract intentionally stays small.  A Report
keeps the graph objects needed to render a cited document in this separate
module so a future persistent adapter can replace the in-memory repository
without changing the route or renderer boundary.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from datetime import datetime
from enum import StrEnum
from typing import Any, Protocol, TypeAlias, runtime_checkable

from pydantic import AliasChoices, BaseModel, ConfigDict, Field


REPORT_ID_PATTERN = r"^rpt_[A-Za-z0-9_-]+$"
STATEMENT_ID_PATTERN = r"^stmt_[A-Za-z0-9_-]+$"
CLAIM_ID_PATTERN = r"^clm_[A-Za-z0-9_-]+$"
EVIDENCE_ID_PATTERN = r"^ev_[A-Za-z0-9_-]+$"
SOURCE_ID_PATTERN = r"^src_[A-Za-z0-9_-]+$"
RESEARCH_RUN_ID_PATTERN = r"^rr_[A-Za-z0-9_-]+$"
RELATION_ID_PATTERN = r"^rel_[A-Za-z0-9_-]+$"
SHA256_PATTERN = r"^sha256:[a-f0-9]{64}$"


class ReportStatus(StrEnum):
    """Lifecycle values from the shared Report contract."""

    DRAFT = "draft"
    READY = "ready"
    PUBLISHED = "published"
    FAILED = "failed"


class ClaimStatus(StrEnum):
    """Claim status values kept in the shared contract."""

    UNSUPPORTED = "unsupported"
    SUPPORTED = "supported"
    REFUTED = "refuted"
    MIXED = "mixed"
    NEEDS_REVIEW = "needs_review"


class EvidenceKind(StrEnum):
    """Evidence kinds from the shared ClaimGraph contract."""

    QUOTE = "quote"
    SUMMARY = "summary"
    TABLE = "table"
    METRIC = "metric"
    OBSERVATION = "observation"


class ClaimEvidenceRelationType(StrEnum):
    """Typed edges between a Claim and an Evidence record."""

    SUPPORTS = "supports"
    REFUTES = "refutes"
    QUALIFIES = "qualifies"
    MENTIONS = "mentions"


class _ReportModel(BaseModel):
    """Shared validation settings for Report graph records."""

    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
    )


class ReportSource(_ReportModel):
    """Source metadata plus an optional original-content snapshot.

    The canonical artifact schema does not store source text.  The API report
    snapshot may receive it as ``original_content`` so the renderer can keep it
    visibly separate from Evidence's generated quote or summary.
    """

    source_id: str = Field(pattern=SOURCE_ID_PATTERN)
    research_run_id: str = Field(pattern=RESEARCH_RUN_ID_PATTERN)
    title: str = Field(min_length=1)
    locator: str = Field(min_length=1)
    original_content: str | None = Field(
        default=None,
        validation_alias=AliasChoices("original_content", "content"),
    )
    source_type: str | None = None
    publisher: str | None = Field(default=None, min_length=1)
    published_at: datetime | None = None
    retrieved_at: datetime | None = None
    content_hash: str | None = Field(default=None, pattern=SHA256_PATTERN)
    metadata: dict[str, str | int | float | bool | None] = Field(
        default_factory=dict
    )

    @property
    def content(self) -> str | None:
        """Compatibility accessor for callers that use ``content``."""

        return self.original_content


class ReportEvidence(_ReportModel):
    """An Evidence item linked to a Source."""

    evidence_id: str = Field(pattern=EVIDENCE_ID_PATTERN)
    source_id: str = Field(pattern=SOURCE_ID_PATTERN)
    research_run_id: str = Field(pattern=RESEARCH_RUN_ID_PATTERN)
    kind: EvidenceKind = EvidenceKind.SUMMARY
    quote_or_summary: str = Field(
        min_length=1,
        validation_alias=AliasChoices("quote_or_summary", "summary"),
    )
    locator: str | None = Field(default=None, min_length=1)
    content_hash: str | None = Field(default=None, pattern=SHA256_PATTERN)
    collected_at: datetime | None = None
    metadata: dict[str, str | int | float | bool | None] = Field(
        default_factory=dict
    )


class ReportClaim(_ReportModel):
    """A Claim rendered as one or more cited report statements."""

    claim_id: str = Field(pattern=CLAIM_ID_PATTERN)
    research_run_id: str = Field(pattern=RESEARCH_RUN_ID_PATTERN)
    text: str = Field(min_length=1)
    status: ClaimStatus
    confidence: float | None = Field(default=None, ge=0, le=1)
    created_at: datetime | None = None
    unsupported_reason: str | None = Field(
        default=None,
        min_length=1,
        validation_alias=AliasChoices("unsupported_reason", "reason"),
    )
    metadata: dict[str, str | int | float | bool | None] = Field(
        default_factory=dict
    )

    @property
    def reason(self) -> str | None:
        """Short alias for the explicit unsupported reason."""

        return self.unsupported_reason


class ClaimEvidenceRelation(_ReportModel):
    """A typed Claim-to-Evidence graph edge."""

    relation_id: str | None = Field(default=None, pattern=RELATION_ID_PATTERN)
    claim_id: str = Field(pattern=CLAIM_ID_PATTERN)
    evidence_id: str = Field(pattern=EVIDENCE_ID_PATTERN)
    relation_type: ClaimEvidenceRelationType = ClaimEvidenceRelationType.SUPPORTS
    confidence: float | None = Field(default=None, ge=0, le=1)
    created_at: datetime | None = None
    metadata: dict[str, str | int | float | bool | None] = Field(
        default_factory=dict
    )


class ReportStatement(_ReportModel):
    """An ordered report sentence connected to exactly one Claim."""

    statement_id: str = Field(default="stmt_generated", pattern=STATEMENT_ID_PATTERN)
    report_id: str | None = Field(default=None, pattern=REPORT_ID_PATTERN)
    claim_id: str = Field(pattern=CLAIM_ID_PATTERN)
    position: int = Field(default=0, ge=0)
    text: str | None = Field(default=None, min_length=1)
    created_at: datetime | None = None
    metadata: dict[str, str | int | float | bool | None] = Field(
        default_factory=dict
    )


class Report(_ReportModel):
    """Stored Report graph snapshot used by the renderer and API route."""

    report_id: str = Field(pattern=REPORT_ID_PATTERN)
    research_run_id: str = Field(pattern=RESEARCH_RUN_ID_PATTERN)
    title: str = Field(min_length=1)
    status: ReportStatus
    created_at: datetime
    published_at: datetime | None = None
    statements: list[ReportStatement] = Field(default_factory=list)
    claims: list[ReportClaim] = Field(default_factory=list)
    evidence: list[ReportEvidence] = Field(default_factory=list)
    sources: list[ReportSource] = Field(default_factory=list)
    claim_evidence_relations: list[ClaimEvidenceRelation] = Field(
        default_factory=list
    )
    metadata: dict[str, str | int | float | bool | None] = Field(
        default_factory=dict
    )


class ReportResponse(_ReportModel):
    """Minimum HTTP response for a Markdown Report lookup."""

    report_id: str = Field(pattern=REPORT_ID_PATTERN)
    research_run_id: str = Field(pattern=RESEARCH_RUN_ID_PATTERN)
    title: str = Field(min_length=1)
    status: ReportStatus
    created_at: datetime
    published_at: datetime | None = None
    markdown: str = Field(min_length=1)

    @classmethod
    def from_report(cls, report: Report) -> "ReportResponse":
        """Build the public response and render Markdown at the boundary."""

        from app.reporting.markdown import render_markdown

        return cls(
            report_id=report.report_id,
            research_run_id=report.research_run_id,
            title=report.title,
            status=report.status,
            created_at=report.created_at,
            published_at=report.published_at,
            markdown=render_markdown(report),
        )


ReportInput: TypeAlias = Report | Mapping[str, Any]


@runtime_checkable
class ReportRepository(Protocol):
    """Storage boundary required by the Report route."""

    async def save(self, report: Report) -> Report:
        """Persist and return a Report."""

    async def get_by_research_run_id(self, research_run_id: str) -> Report | None:
        """Return the Report for a ResearchRun, if one exists."""


class InMemoryReportRepository:
    """Deterministic repository adapter for local development and tests."""

    def __init__(self, reports: Iterable[ReportInput] | Mapping[str, ReportInput] = ()) -> None:
        self._reports: dict[str, Report] = {}
        self._report_ids_by_research_run: dict[str, str] = {}

        if isinstance(reports, Mapping) and "report_id" not in reports:
            report_values = reports.values()
        elif isinstance(reports, Mapping):
            report_values = (reports,)
        else:
            report_values = reports

        for report in report_values:
            self.put(report)

    def put(self, report: ReportInput) -> Report:
        """Validate and synchronously store a Report for test setup."""

        record = report if isinstance(report, Report) else Report.model_validate(report)
        stored = record.model_copy(deep=True)
        previous_report_id = self._report_ids_by_research_run.get(
            stored.research_run_id
        )
        if previous_report_id is not None and previous_report_id != stored.report_id:
            self._reports.pop(previous_report_id, None)
        self._reports[stored.report_id] = stored
        self._report_ids_by_research_run[stored.research_run_id] = stored.report_id
        return stored.model_copy(deep=True)

    add = put

    async def save(self, report: ReportInput) -> Report:
        """Persist a Report through the async repository boundary."""

        return self.put(report)

    async def create(self, report: ReportInput) -> Report:
        """Alias matching the existing ResearchRun repository vocabulary."""

        return await self.save(report)

    async def get_by_research_run_id(self, research_run_id: str) -> Report | None:
        report_id = self._report_ids_by_research_run.get(research_run_id)
        report = self._reports.get(report_id) if report_id is not None else None
        return report.model_copy(deep=True) if report is not None else None

    async def get_by_id(self, report_id: str) -> Report | None:
        report = self._reports.get(report_id)
        return report.model_copy(deep=True) if report is not None else None

    async def get(self, report_id: str) -> Report | None:
        """Compatibility alias for repository adapters using ``get``."""

        return await self.get_by_id(report_id)

    def __len__(self) -> int:
        return len(self._reports)


# Common domain spellings keep the small contract easy to discover without
# duplicating model definitions.
Claim = ReportClaim
Evidence = ReportEvidence
Source = ReportSource
ReportRecord = Report
ReportModel = Report


__all__ = [
    "Claim",
    "ClaimEvidenceRelation",
    "ClaimEvidenceRelationType",
    "ClaimStatus",
    "Evidence",
    "EvidenceKind",
    "InMemoryReportRepository",
    "Report",
    "ReportClaim",
    "ReportEvidence",
    "ReportInput",
    "ReportModel",
    "ReportRecord",
    "ReportRepository",
    "ReportResponse",
    "ReportSource",
    "ReportStatement",
    "ReportStatus",
    "Source",
]

File: apps/web/app/lib/report.ts
export type SourceType = "web" | "document" | "paper" | "dataset" | "api" | "manual";
export type EvidenceKind = "quote" | "summary" | "table" | "metric" | "observation";
export type ClaimStatus = "unsupported" | "supported" | "refuted" | "mixed" | "needs_review";
export type ClaimEvidenceRelationType = "supports" | "refutes" | "qualifies" | "mentions";
export type ReportStatus = "draft" | "ready" | "published" | "failed";

export type Source = {
  source_id: string;
  research_run_id: string;
  source_type: SourceType;
  title: string;
  locator: string;
  source_url: string | null;
  publisher?: string;
  retrieved_at: string;
  content_hash: string;
};

export type Evidence = {
  evidence_id: string;
  source_id: string;
  research_run_id: string;
  kind: EvidenceKind;
  quote_or_summary: string;
  locator: string;
  collected_at: string;
  content_hash: string;
};

export type Claim = {
  claim_id: string;
  research_run_id: string;
  text: string;
  status: ClaimStatus;
  confidence: number;
  created_at: string;
};

export type ClaimEvidenceRelation = {
  relation_id: string;
  claim_id: string;
  evidence_id: string;
  relation_type: ClaimEvidenceRelationType;
  confidence: number;
  created_at: string;
};

export type ReportStatement = {
  statement_id: string;
  report_id: string;
  claim_id: string;
  position: number;
  text: string;
  created_at: string;
};

export type Report = {
  report_id: string;
  research_run_id: string;
  question: string;
  title: string;
  status: ReportStatus;
  markdown: string;
  created_at: string;
  statements: ReportStatement[];
  claims: Claim[];
  evidence: Evidence[];
  sources: Source[];
  claim_evidence_relations: ClaimEvidenceRelation[];
};

export type ReportResponse = {
  report: Report | null;
};

const SOURCE_TYPES = ["web", "document", "paper", "dataset", "api", "manual"] as const;
const EVIDENCE_KINDS = ["quote", "summary", "table", "metric", "observation"] as const;
const CLAIM_STATUSES = [
  "unsupported",
  "supported",
  "refuted",
  "mixed",
  "needs_review",
] as const;
const RELATION_TYPES = ["supports", "refutes", "qualifies", "mentions"] as const;
const REPORT_STATUSES = ["draft", "ready", "published", "failed"] as const;
const SHA256_PATTERN = /^sha256:[a-f0-9]{64}$/;

export function parseReportResponse(value: unknown): ReportResponse {
  const response = requireRecord(value, "ReportResponse");
  if (!("report" in response)) {
    throw new TypeError("ReportResponse.report is required");
  }

  return {
    report: response.report === null ? null : parseReport(response.report),
  };
}

function parseReport(value: unknown): Report {
  const record = requireRecord(value, "Report");
  const report: Report = {
    report_id: requireString(record.report_id, "Report.report_id"),
    research_run_id: requireString(record.research_run_id, "Report.research_run_id"),
    question: requireString(record.question, "Report.question"),
    title: requireString(record.title, "Report.title"),
    status: requireEnum(record.status, REPORT_STATUSES, "Report.status"),
    markdown: requireString(record.markdown, "Report.markdown"),
    created_at: requireString(record.created_at, "Report.created_at"),
    statements: requireArray(record.statements, "Report.statements").map(parseStatement),
    claims: requireArray(record.claims, "Report.claims").map(parseClaim),
    evidence: requireArray(record.evidence, "Report.evidence").map(parseEvidence),
    sources: requireArray(record.sources, "Report.sources").map(parseSource),
    claim_evidence_relations: requireArray(
      record.claim_evidence_relations,
      "Report.claim_evidence_relations",
    ).map(parseRelation),
  };

  if (report.statements.length === 0) {
    throw new TypeError("Report.statements must not be empty");
  }
  validateReportGraph(report);
  return report;
}

function parseSource(value: unknown): Source {
  const record = requireRecord(value, "Source");
  const sourceUrl = "source_url" in record
    ? record.source_url === null
      ? null
      : requireString(record.source_url, "Source.source_url")
    : null;

  if (sourceUrl !== null && !/^https:\/\//i.test(sourceUrl)) {
    throw new TypeError("Source.source_url must use HTTPS");
  }

  return {
    source_id: requireString(record.source_id, "Source.source_id"),
    research_run_id: requireString(record.research_run_id, "Source.research_run_id"),
    source_type: requireEnum(record.source_type, SOURCE_TYPES, "Source.source_type"),
    title: requireString(record.title, "Source.title"),
    locator: requireString(record.locator, "Source.locator"),
    source_url: sourceUrl,
    publisher: optionalString(record.publisher, "Source.publisher"),
    retrieved_at: requireString(record.retrieved_at, "Source.retrieved_at"),
    content_hash: requireHash(record.content_hash, "Source.content_hash"),
  };
}

function parseEvidence(value: unknown): Evidence {
  const record = requireRecord(value, "Evidence");
  return {
    evidence_id: requireString(record.evidence_id, "Evidence.evidence_id"),
    source_id: requireString(record.source_id, "Evidence.source_id"),
    research_run_id: requireString(record.research_run_id, "Evidence.research_run_id"),
    kind: requireEnum(record.kind, EVIDENCE_KINDS, "Evidence.kind"),
    quote_or_summary: requireString(record.quote_or_summary, "Evidence.quote_or_summary"),
    locator: requireString(record.locator, "Evidence.locator"),
    collected_at: requireString(record.collected_at, "Evidence.collected_at"),
    content_hash: requireHash(record.content_hash, "Evidence.content_hash"),
  };
}

function parseClaim(value: unknown): Claim {
  const record = requireRecord(value, "Claim");
  return {
    claim_id: requireString(record.claim_id, "Claim.claim_id"),
    research_run_id: requireString(record.research_run_id, "Claim.research_run_id"),
    text: requireString(record.text, "Claim.text"),
    status: requireEnum(record.status, CLAIM_STATUSES, "Claim.status"),
    confidence: requireConfidence(record.confidence, "Claim.confidence"),
    created_at: requireString(record.created_at, "Claim.created_at"),
  };
}

function parseRelation(value: unknown): ClaimEvidenceRelation {
  const record = requireRecord(value, "ClaimEvidenceRelation");
  return {
    relation_id: requireString(record.relation_id, "ClaimEvidenceRelation.relation_id"),
    claim_id: requireString(record.claim_id, "ClaimEvidenceRelation.claim_id"),
    evidence_id: requireString(record.evidence_id, "ClaimEvidenceRelation.evidence_id"),
    relation_type: requireEnum(
      record.relation_type,
      RELATION_TYPES,
      "ClaimEvidenceRelation.relation_type",
    ),
    confidence: requireConfidence(
      record.confidence,
      "ClaimEvidenceRelation.confidence",
    ),
    created_at: requireString(record.created_at, "ClaimEvidenceRelation.created_at"),
  };
}

function parseStatement(value: unknown): ReportStatement {
  const record = requireRecord(value, "ReportStatement");
  return {
    statement_id: requireString(record.statement_id, "ReportStatement.statement_id"),
    report_id: requireString(record.report_id, "ReportStatement.report_id"),
    claim_id: requireString(record.claim_id, "ReportStatement.claim_id"),
    position: requireInteger(record.position, "ReportStatement.position"),
    text: requireString(record.text, "ReportStatement.text"),
    created_at: requireString(record.created_at, "ReportStatement.created_at"),
  };
}

function validateReportGraph(report: Report): void {
  const reportStatementIds = uniqueIds(
    report.statements.map((statement) => statement.statement_id),
    "Report statements",
  );
  const claimIds = uniqueIds(
    report.claims.map((claim) => claim.claim_id),
    "Claim",
  );
  const evidenceIds = uniqueIds(
    report.evidence.map((evidence) => evidence.evidence_id),
    "Evidence",
  );
  const sourceIds = uniqueIds(
    report.sources.map((source) => source.source_id),
    "Source",
  );
  uniqueIds(
    report.claim_evidence_relations.map((relation) => relation.relation_id),
    "ClaimEvidenceRelation",
  );

  if (reportStatementIds.size !== report.statements.length) {
    throw new TypeError("Report statements must have unique IDs");
  }

  for (const statement of report.statements) {
    if (statement.report_id !== report.report_id) {
      throw new TypeError(`ReportStatement ${statement.statement_id} has the wrong report ID`);
    }
    if (!claimIds.has(statement.claim_id)) {
      throw new TypeError(`ReportStatement ${statement.statement_id} references a missing Claim`);
    }
  }

  for (const claim of report.claims) {
    if (claim.research_run_id !== report.research_run_id) {
      throw new TypeError(`Claim ${claim.claim_id} has the wrong ResearchRun ID`);
    }
  }

  for (const source of report.sources) {
    if (source.research_run_id !== report.research_run_id) {
      throw new TypeError(`Source ${source.source_id} has the wrong ResearchRun ID`);
    }
  }

  for (const evidence of report.evidence) {
    if (evidence.research_run_id !== report.research_run_id) {
      throw new TypeError(`Evidence ${evidence.evidence_id} has the wrong ResearchRun ID`);
    }
    if (!sourceIds.has(evidence.source_id)) {
      throw new TypeError(`Evidence ${evidence.evidence_id} references a missing Source`);
    }
  }

  const relationsByClaim = new Map<string, ClaimEvidenceRelation[]>();
  for (const relation of report.claim_evidence_relations) {
    if (!claimIds.has(relation.claim_id)) {
      throw new TypeError(`Relation ${relation.relation_id} references a missing Claim`);
    }
    if (!evidenceIds.has(relation.evidence_id)) {
      throw new TypeError(`Relation ${relation.relation_id} references missing Evidence`);
    }
    const relations = relationsByClaim.get(relation.claim_id) ?? [];
    relationsByClaim.set(relation.claim_id, [...relations, relation]);
  }

  for (const claim of report.claims) {
    const relations = relationsByClaim.get(claim.claim_id) ?? [];
    if (claim.status === "unsupported") {
      continue;
    }
    if (relations.length === 0) {
      throw new TypeError(`Claim ${claim.claim_id} requires an Evidence relation`);
    }
    if (claim.status === "supported" && !relations.some((item) => item.relation_type === "supports")) {
      throw new TypeError(`Claim ${claim.claim_id} requires a supports relation`);
    }
    if (claim.status === "refuted" && !relations.some((item) => item.relation_type === "refutes")) {
      throw new TypeError(`Claim ${claim.claim_id} requires a refutes relation`);
    }
    if (
      claim.status === "mixed" &&
      (!relations.some((item) => item.relation_type === "supports") ||
        !relations.some((item) => item.relation_type === "refutes"))
    ) {
      throw new TypeError(`Claim ${claim.claim_id} requires both supports and refutes relations`);
    }
  }
}

function uniqueIds(values: string[], label: string): Set<string> {
  const ids = new Set(values);
  if (ids.size !== values.length) {
    throw new TypeError(`${label} IDs must be unique`);
  }
  return ids;
}

function requireRecord(value: unknown, field: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function requireArray(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) {
    throw new TypeError(`${field} must be an array`);
  }
  return value;
}

function requireString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new TypeError(`${field} must be a non-empty string`);
  }
  return value;
}

function optionalString(value: unknown, field: string): string | undefined {
  if (value === undefined) {
    return undefined;
  }
  return requireString(value, field);
}

function requireEnum<const Values extends readonly string[]>(
  value: unknown,
  values: Values,
  field: string,
): Values[number] {
  if (typeof value !== "string" || !values.includes(value)) {
    throw new TypeError(`${field} has an unsupported value`);
  }
  return value as Values[number];
}

function requireConfidence(value: unknown, field: string): number {
  if (typeof value !== "number" || !Number.isFinite(value) || value < 0 || value > 1) {
    throw new TypeError(`${field} must be a number from 0 to 1`);
  }
  return value;
}

function requireInteger(value: unknown, field: string): number {
  if (typeof value !== "number" || !Number.isInteger(value) || value < 0) {
    throw new TypeError(`${field} must be a non-negative integer`);
  }
  return value;
}

function requireHash(value: unknown, field: string): string {
  const hash = requireString(value, field);
  if (!SHA256_PATTERN.test(hash)) {
    throw new TypeError(`${field} must be a sha256 hash`);
  }
  return hash;
}
